motoko/invoice-canister/examples/motoko-seller-client/backend/src/mocks/invoice.mo

```diff
@@ -0,0 +1,608 @@
+      };
+      case(_){
+        return errInvalidToken;
+      };
+    };
+  };
+// #endregion
+// #endregion
+
+// #region Get Invoice
+  public shared query ({caller}) func get_invoice (args: T.GetInvoiceArgs) : async T.GetInvoiceResult {
+    let invoice = invoices.get(args.id);
+    switch(invoice){
+      case(null){
+        return #err({
+          message = ?"Invoice not found";
+          kind = #NotFound;
+        });
+      };
+      case (?i) {
+        if (i.creator == caller) {
+          return #ok({invoice = i});
+        };
+        // If additional permissions are provided
+        switch (i.permissions) {
+          case (null) {
+            return #err({
+              message = ?"You do not have permission to view this invoice";
+              kind = #NotAuthorized;
+            });
+          };
+          case (?permissions) {
+            let hasPermission = Array.find<Principal>(
+              permissions.canGet,
+              func (x : Principal) : Bool {
+                return x == caller;
+              }
+            );
+            if (Option.isSome(hasPermission)) {
+              return #ok({invoice = i});
+            } else {
+              return #err({
+                message = ?"You do not have permission to view this invoice";
+                kind = #NotAuthorized;
+              });
+            };
+          };
+        };
+        #ok({invoice = i});
+      };
+    };
+  };
+// #endregion
+
+// #region Get Balance
+  public shared ({caller}) func get_balance (args: T.GetBalanceArgs) : async T.GetBalanceResult {
+    let token = args.token;
+    let canisterId = Principal.fromActor(InvoiceMock);
+    switch(token.symbol){
+      case("ICP"){
+        let defaultAccount = U.getDefaultAccount({
+          canisterId;
+          principal = caller;
+        });
+        let balance = icpLedgerMock.get(defaultAccount);
+        switch(balance){
+          case(null){
+            return #err({
+              message = ?"Could not get balance";
+              kind = #NotFound;
+            });
+          };
+          case(? b){
+            return #ok({balance = b});
+          };
+        };
+      };
+      case(_){
+        return errInvalidToken;
+      };
+    };
+  };
+// #endregion
+
+// #region Verify Invoice
+  public shared ({caller}) func verify_invoice (args: T.VerifyInvoiceArgs) : async T.VerifyInvoiceResult {
+    let invoice = invoices.get(args.id);
+    let canisterId = Principal.fromActor(InvoiceMock);
+
+    switch(invoice){
+      case(null){
+        return #err({
+          message = ?"Invoice not found";
+          kind = #NotFound;
+        });
+      };
+      case(? i){
+        // Return if already verified
+        if (i.verifiedAtTime != null){
+          return #ok(#AlreadyVerified{
+            invoice = i;
+          });
+        };
+
+        if (i.creator != caller) {
+          switch (i.permissions) {
+            case null {
+              return #err({
```

Comment:
Is it expected that someone without permission to verify an invoice (this check) can find out that it's already been verified (the above check)?

Reply:
I'll create a follow-up issue for this